src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,286 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
+                if (!navPropNames.Contains(prop.Name))
                 {
-                    IODataAPIHandler apiHandlerNested = this.GetNestedHandler(this.ParentObject, odataPath.LastSegment.Identifier);
-                    object[] getParams = new object[] { keys, null, null };
-                    if (apiHandlerNested.GetType().GetMethod(nameof(TryGet)).Invoke(apiHandlerNested, getParams).Equals(ODataAPIResponseStatus.Success))
+                    object resVal = prop.GetValue(originalObj);
+
+                    if (resVal != null)
                     {
-                        return getParams[1];
+                        prop.SetValue(newObject, resVal);
                     }
-                    else
+                }
+            }
+
+            return newObject;
+        }
+
+        private static void CopyProperties(object originalObject, object newObject, List<string> navPropNames)
+        {
+            foreach (PropertyInfo prop in originalObject.GetType().GetProperties())
+            {
+                // Don't copy Navigation Properties. They will be handled in the next level nesting.
+                if (!navPropNames.Contains(prop.Name))
+                {
+                    object resVal = prop.GetValue(originalObject);
+
+                    if (resVal != null)
                     {
-                        if (apiHandlerNested.GetType().GetMethod(nameof(TryCreate)).Invoke(apiHandlerNested, getParams).Equals(ODataAPIResponseStatus.Success))
-                        {
-                            return getParams[1];
-                        }
-                        else
-                        {
-                            return null;
-                        }
+                        prop.SetValue(newObject, resVal);
                     }
                 }
             }
+        }
+
+        private static IDictionary<string, object> GetKeys(IEnumerable<IEdmStructuralProperty> properties, object resource, Type type)
+        {
+            IDictionary<string, object> keys = new Dictionary<string, object>();
+
+            foreach (IEdmStructuralProperty property in properties)
+            {
+                PropertyInfo prop = type.GetProperty(property.Name);
+                object value = prop.GetValue(resource, null);
+
+                keys.Add(new KeyValuePair<string, object>(property.Name, value));
+            }
+
+            return keys;
+        }
 
-            return null;
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We don't need to throw an exception but instead return odataPath as null.")]
+        private static ODataPath GetODataPath(string path, IEdmModel model)
+        {
+            ODataPath odataPath;
+            try
+            {
+                ODataUriParser parser = new ODataUriParser(model, new Uri(path, UriKind.Relative));
```

Comment:
For the `UriParser`, there is no need. What would be needed is the `ODataQueryOptionsParser` because of configurations like `NoDollarQueryOptions`.

Other areas in the codebase e.g `ODataResourceDeserializer` still initializes a new `ODataUriParser`

Reply:
Wont fix for now